### PR TITLE
HDF5 storage layout diagrams

### LINKED ISSUE #124: DOC: Pseudo UML description for the hdf5 serialization layout

### PR DESCRIPTION
**Pull request by roigcarlo**
This PR fixes #124.

I am not sure if I fully understood the lattice scheme, so please take a loot at that

### CODE REVIEW

**Comment by tuopuu** [comment]
@roigcarlo, pictures look nice, but I don't think h5_lattice diagram is quite correct, although I might have understood the diagrams wrong. I'll take a better look tomorrow, but here are some problems that I noticed: 
* group.attrs.size, group.attrs.origin etc are not shown anywhere. I think they should be included under H5Lattice :: Group, for example like group.attrs.size = size.
* Lattice::Group and Data::Group should be typed as table instead of group, because they are directly under H5Lattice-group (as "files", not as "subfolders").

**Comment by itziakos** [comment]
@roigcarlo a number of comments on the H5Mesh:

- It looks that the H5Mesh layout is missing the `data` Table.
- `points_uids` is an array of strings not a Table
- I suggest to add the string size in brackets (e.g string[32]) to provide the size of the column types

We should also rename points_uids -> point_uids at some point :smile:

**Comment by roigcarlo** [comment]
Hi @itziakos,  @tuopuu, @nathanfranklin 

I have changed H5Mesh based on your comments.

Still I am uncertain of what to do with lattice since it seems that it does not follow the schema of Particles and Mesh. Tuomas has explained to me how lattice was working, and he has made some changes that we have uploaded in https://github.com/simphony/simphony-common/commit/cd879a4e18f3d46a1e5482e7ee942efd902bc8cb,.

In general I am not sure how to deal with the attributes of lattice, for instance "attrs.type". I strongly believe that this cannot appear in a UML so the ideas are either to put them simply to the lattice class or  
separate them into a different class.

Digging a little bit about the first I found that Mesh and Particles are actually modelled after a PyTables approach ( for example see http://www.pytables.org/usersguide/introduction.html ), so I believe that this should be the way, but I wanted to discuss it first here.

What do you think?

**Comment by itziakos** [comment]
> In general I am not sure how to deal with the attributes of lattice, for instance "attrs.type". I strongly believe that this cannot appear in a UML so the ideas are either to put them simply to the lattice class or

Yes as you point later the HDF5 layout is based on a pseudo uml description (this is how pytables describes it). The aim is to show the **layout** of the stored data using Pytables/HDF5 entities.

File: doc/source/uml/h5mesh.txt [hunk 1]
```diff
@@ -0,0 +1,67 @@
...
+class Mesh as "H5Mesh::Group" {
+  _v_name: String = "File"
+  points : Array of String[32]
+  edges : Edge
+  faces : Face
+  cells : Cell
+  data : IndexedDataContainer
+}
+
+class Point as "Point::Group" {
```
Comment (itziakos) [inline doc/source/uml/h5mesh.txt]: Point::Group -> Point::Table

Comment (itziakos) [inline doc/source/uml/h5mesh.txt]: More appropriate I think we should declare
```
class Mesh as "H5Mesh(Group)" {
+  _v_name: String = <name>
+  points : Table = Points 
+  edges : Table = Edges 
+  faces : Table = Faces
+  cells : Table = Faces
+  data : Table = Data
+}
```

File: doc/source/uml/h5lattice.txt [hunk 2]
```diff
@@ -0,0 +1,37 @@
+@startuml
+class H5Lattice as "H5Lattice::Group" {
+  _v_name: String = name
+  _lattice: Table
+  _data: Table
+}
+
+class Lattice as "_table::Table" {
+  _v_name: String = "lattice"
+  attrs.type: String = type
```
Comment (itziakos) [inline doc/source/uml/h5lattice.txt]: Attributes can be placed inside a group like this
The Lattice class can be described as follows

```
class H5Lattice as "H5Lattice(Group)" {
  _v_name: String = <name>
  lattice : Table = Data
  data : Table = Data
  -- Node Attributes --
  type : 
  base_vect : Double[2..3]
  size : Int[2..3]
  origin : Double[2..3]

class Data as "Data(Table)" {
  _v_name: String = <name>
  Row : DataRow
  uid : String[32]
  data: CUBARow
  mask: Bool[size of CUBARow]
}

class DataRow {
    
}
 ```

**Comment by itziakos** [comment]
Hi Guys, 
I have added some example on how I think we should go about providing a diagram of the hdf5 layout based on pytables entities (not the various wrapper classes). This is not pure UML but it makes clear to the reader that this is how the data are stored in the hdf5. IIn some cases we have to add some notes and correct but it does not have to be perfect at this iteration.

**Comment by roigcarlo** [comment]
Hi I implemented all the changes, is that ok?

File: doc/source/uml/h5lattice.txt [hunk 3]
```diff
@@ -0,0 +1,45 @@
...
+  base_vect : Double[2..3]
+  size : Int[2..3]
+  origin: Double[2..3]
+}
+
+class Nodes as "Nodes(Table)" {
+  _v_name: String = "lattice"
+  row : Record
+}
+note left : index into the lattice node table is calculated using numpy....
```
Comment (itziakos) [inline doc/source/uml/h5lattice.txt]: This should be expanded (this is why I placed the ...) :smile:

File: doc/source/uml/h5lattice.txt [hunk 4]
```diff
@@ -0,0 +1,45 @@
...
+
+class Data as "Data(Table)" {
+  _v_name: String = "data"
+  row : NoUidRecord
+}
+
+class Record {
+  [ uid ]: String[32]
+  [ data ]: DataRow
+  [ mask ]: Bool[size of CUBARow]
```
Comment (itziakos) [inline doc/source/uml/h5lattice.txt]: replace `CUBARow` -> `DataRow`

File: doc/source/uml/h5particles.txt [hunk 5]
```diff
@@ -0,0 +1,64 @@
...
+
+class ParticleTableRow as "Particle(Row)" {
+   [ uid ] : String[32]
+   [ coordinates ] : Float64[3]
+}
+
+class BondTableRow as "Bond(Row)" {
+   [ uid ] : String[32]
+   [ particles ] : Integer
+   [ n_particles ] : Integer = MAX_NUMBER_PARTICLES_IN_BOND
```
Comment (itziakos) [inline doc/source/uml/h5particles.txt]: I would probably have this as `0 <= Integer <= MAX_NUMBER_PARTICLES_IN_BOND` to show the range better

File: doc/source/uml/h5particles.txt [hunk 6]
```diff
@@ -0,0 +1,64 @@
...
+}
+
+class DataRow as "Data(Row)" {
+  [ uid ]: String[32]
+  [ data ]: CUBARow
```
Comment (itziakos) [inline doc/source/uml/h5particles.txt]: This probably should be `Record` for consistancy

File: doc/source/uml/h5particles.txt [hunk 7]
```diff
@@ -0,0 +1,64 @@
...
+  [ mask ]: Bool[size of CUBARow]
```
Comment (itziakos) [inline doc/source/uml/h5particles.txt]: Similar to other places, this should be Record

File: doc/source/uml/h5mesh.txt [hunk 8]
```diff
@@ -0,0 +1,81 @@
...
+  [ data ] : String[32]
+  [ point_uids ] : Array of String[32]
+  [ n_points ] : Integer[MAX_POINTS_IN_FACE]
+}
+
+class CellRow as "Cell(Row)" {
+  [ uid ] : String[32]
+  [ data ] : String[32]
+  [ point_uids ] : Array of String[32]
+  [ n_points ] : Integer[MAX_POINTS_IN_CELL]
```
Comment (itziakos) [inline doc/source/uml/h5mesh.txt]: see https://github.com/simphony/simphony-common/pull/142/files#r29815437

File: doc/source/uml/h5mesh.txt [hunk 9]
```diff
@@ -0,0 +1,81 @@
...
+  _v_name : string = "Cells"
+ row: CellRow
+}
+
+class Data as "Data(Table)" {
+  _v_name: String = "Data"
+  row : DataRow
+}
+
+class PointRow as "Points(Row)" {
```
Comment (itziakos) [inline doc/source/uml/h5mesh.txt]: Points(Row) -> PointRow

File: doc/source/uml/h5mesh.txt [hunk 10]
```diff
@@ -0,0 +1,81 @@
...
+class CellRow as "Cell(Row)" {
+  [ uid ] : String[32]
+  [ data ] : String[32]
+  [ point_uids ] : Array of String[32]
+  [ n_points ] : Integer[MAX_POINTS_IN_CELL]
+}
+
+class DataRow as "Data(Row)" {
+  [ uid ]: String[32]
+  [ data ]: CUBARow
```
Comment (itziakos) [inline doc/source/uml/h5mesh.txt]: see https://github.com/simphony/simphony-common/pull/142/files#r29815532

File: doc/source/uml/h5particles.txt (hunk 6 above)
Comment (itziakos) [inline doc/source/uml/h5particles.txt]: We probably need to revisit the names so that we are consistent

**Comment by itziakos** [comment]
@roigcarlo, regarding the umls there just a couple of small thinks and make sure that we are consistent. We should probably factor out the common components into separate files similar to what we did for the api. Finally we need to add the png files into an rst file (a little test will also be nice) and make it part of the documentation index (they will not be part of the documentation otherwise).

I can make the changes today (Thursday morning) if that is ok with you.

**Comment by roigcarlo** [comment]
@itziakos Yes sure

**Comment by itziakos** [comment]
@roigcarlo , @tuopuu and @nathanfranklin, I have updated the diagrams and added the HDF5 storage page to the documentation. Please have another look

File: doc/source/hdf5_storage.rst [hunk 11]
```diff
@@ -0,0 +1,60 @@
...
+manipulating CUDS containers. Please also note that returned containers from the get
+methods are live proxy objects on top of the HDF5 storage (in contrast to the common
+offline save and read operations).
+
+
+HDF5 Stored Layout
+------------------
+
+Data are stored in HDF5 files using a separate layout for each type of CUDS container.
+The stored layout of the containers is provided bellow using a pseudo-uml description
```
Comment (nathanfranklin) [inline doc/source/hdf5_storage.rst]: bellow -> below

File: doc/source/hdf5_storage.rst [hunk 12]
```diff
@@ -0,0 +1,60 @@
...
+
+.. rubric:: Particles
+
+.. figure:: ./images/h5particles.png
+
+   **Figure 2:** Diagram of the Particles based storage.
+
+   The Particles container is stored using one table for the container
+   `data` attribute and two groups to holding the particle and bond
+   items saperatly. Each item group has two tables one for the item
```
Comment (nathanfranklin) [inline doc/source/hdf5_storage.rst]: saperatly.

File: doc/source/hdf5_storage.rst [hunk 13]
```diff
@@ -0,0 +1,60 @@
+HDF5 Storage
+============
+
+Cuds containers can be stored in HDF5 files using the :class:`~.H5File` class. The provided
+api is a currently a reduced version of the Modeling Engine api and supports adding and
```
Comment (roigcarlo) [inline doc/source/hdf5_storage.rst]: is a currently a -> is currently a

File: doc/source/hdf5_storage.rst [hunk 14]
```diff
@@ -0,0 +1,60 @@
...
+HDF5 Stored Layout
+------------------
+
+Data are stored in HDF5 files using a separate layout for each type of CUDS container.
+The stored layout of the containers is provided bellow using a pseudo-uml description
+for the HDF5 based layout of the data stored in the files.
+
+.. warning::
+
+   This is the provisional storage layout and is under continues development. Backwards
```
Comment (roigcarlo) [inline doc/source/hdf5_storage.rst]: is under continues development -> is under continuous development

File: doc/source/hdf5_storage.rst [hunk 15]
```diff
@@ -0,0 +1,60 @@
...
+offline save and read operations).
+
+
+HDF5 Stored Layout
+------------------
+
+Data are stored in HDF5 files using a separate layout for each type of CUDS container.
+The stored layout of the containers is provided bellow using a pseudo-uml description
+for the HDF5 based layout of the data stored in the files.
+
```
Comment (nathanfranklin) [inline doc/source/hdf5_storage.rst]: Do we need a figure or text to illustrate they layout of where the individual containers of Mesh, Lattice and Particles are stored?
(``/particle/foo``, `/lattice/bar`..):

**Comment by nathanfranklin** [comment]
I just looked briefly at everything but took a closer look at the `` doc/source/images/h5particles.png `` .  Everything looks good. :+1:   I just left some minor comments (spelling and one question)

File: doc/source/hdf5_storage.rst (hunk 15 above)
Comment (itziakos) [inline doc/source/hdf5_storage.rst]: Good point, I will add a layout for the h5File

Comment (itziakos) [inline doc/source/hdf5_storage.rst]: Done

**Comment by nathanfranklin** [comment]
This can be merged, right?

**Comment by itziakos** [comment]
as far as I can tell

**Pull request merged**